File: Tests/Phase 2/Review1/monotonic-queue/monotonic-queue.py
from typing import List


class MonotonicQueue:

    def __init__(self) -> None:
        self.queue: List[int] = []

    def push(self, value: int):
        while len(self.queue) and self.queue[len(self.queue) - 1] < value:
            self.queue.pop()
        self.queue.append(value)

    def pop(self):
        if len(self.queue):
            self.queue.pop(0)

    def getMax(self):
        return self.queue[0]

    def size(self):
        return len(self.queue)

    def isEmpty(self):
        return len(self.queue) == 0


# test
mq = MonotonicQueue()
mq.push(3)
mq.push(1)
mq.push(4)
mq.push(2)
mq.push(1)
print(mq.queue)

File: Data Structure & Algorithm/Archive/Models/Queue/Monotonic Queue.ts
export class MonotonicQueue {
  queue: number[];

  constructor() {
    this.queue = [];
  }

  push(value: number) {
    while (this.queue.length && this.queue[this.queue.length - 1] < value) {
      this.queue.pop();
    }
    this.queue.push(value);
  }

  pop() {
    if (this.queue.length) {
      this.queue.shift();
    }
  }

  getMax(): number | undefined {
    return this.queue.length ? this.queue[0] : undefined;
  }

  size(): number {
    return this.queue.length;
  }

  isEmpty(): boolean {
    return this.queue.length === 0;
  }
}

// test
const mq = new MonotonicQueue();
mq.push(3);
mq.push(1);
mq.push(4);
mq.push(2);

console.log(mq.getMax()); // 输出: 4
mq.pop();
console.log(mq.getMax()); // 输出: 2
mq.pop();
